internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -21,96 +21,183 @@ import (
+	// name is the name of the secret, and it derives from key
+	name string
+	// namespace is the name of the namespace, and it derives from key
+	namespace string
+	// key is the key inside the secret, and it derives from key
+	key string
+	// value is the value of key inside the secret
+	value string
+	// apiExists is true if the secret was fetched from the API with no errors
+	apiExists bool
+	// apiFetchTime is the time the secret was fetched from the API
+	apiFetchTime time.Time
 }
 
-type secretsData struct {
-	value      string
-	lastAccess time.Time
+type conditionFn func(existing secret, exists bool) bool
+
+type contextProviderK8SSecrets struct {
+	logger    *logger.Logger
+	config    *Config
+	client    k8sclient.Interface
+	clientMtx sync.RWMutex
+	running   chan struct{}
+	store     store
 }
 
-// ContextProviderBuilder builds the context provider.
+// ContextProviderBuilder builds the kubernetes_secrets context provider. By default, this provider employs a cache
+// to reduce the number of requests to the API server. The cache refreshes the secrets referenced during each Fetch call
+// every Config.RefreshInterval. To maintain only secrets that are actually needed by the agent, each secret reference
+// expires based on the Config.TTLDelete. During expiration of secret references or actual changes of secret values,
+// the kubernetes_secrets provider calls the ContextProviderComm.Signal() to notify the agent. The cache mechanism
+// can be disabled by setting Config.DisableCache to true.
 func ContextProviderBuilder(logger *logger.Logger, c *config.Config, _ bool) (corecomp.ContextProvider, error) {
-	var cfg Config
+	cfg := defaultConfig()
+
 	if c == nil {
 		c = config.New()
 	}
-	err := c.UnpackTo(&cfg)
+
+	err := c.UnpackTo(cfg)
 	if err != nil {
 		return nil, errors.New(err, "failed to unpack configuration")
 	}
-	return &contextProviderK8sSecrets{
-		logger:       logger,
-		config:       &cfg,
-		secretsCache: make(map[string]*secretsData),
+	return &contextProviderK8SSecrets{
+		logger:  logger,
+		config:  cfg,
+		client:  nil,
+		running: make(chan struct{}),
+		store:   newExpirationCache(cfg.TTLDelete),
 	}, nil
 }
 
-func (p *contextProviderK8sSecrets) Fetch(key string) (string, bool) {
-	if p.config.DisableCache {
-		valid := p.validateKey(key)
-		if valid {
-			return p.fetchSecretWithTimeout(key)
-		} else {
-			return "", false
-		}
-	} else {
-		return p.getFromCache(key)
-	}
-}
-
 // Run initializes the k8s secrets context provider.
-func (p *contextProviderK8sSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+func (p *contextProviderK8SSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
 	client, err := getK8sClientFunc(p.config.KubeConfig, p.config.KubeClientOptions)
 	if err != nil {
-		p.logger.Debugf("kubernetes_secrets provider skipped, unable to connect: %s", err)
-		return nil
+		// signal that the provider has initialized
+		close(p.running)
```

Comment:
Is this guard necessary? I thought it wasn't legal to call `Fetch` before `Run` returns.

Reply:
In the happy path I see what you say and indeed `Fetch` can't happen before `Run` but if the `Run` is too slow, since this a detached goroutine if I am correct [here](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/composable/controller.go#L392), then a `Fetch` can happen in parallel, right?

Reply:
Allright, makes sense. If this is how we want this to work, we should consider changing `Run` to be non-blocking so we don't need to defensively code in every provider. But that's material for a separate issue.